File: decrypt/response.py
import argparse
import json
from pathlib import Path

from common.protocol import (
    read_base64_file,
    parse_request,
    parse_header_blob1,
    udid_raw_to_canonical_string,
    derive_iv_from_udid_string,
)
from .common import (
    decrypt_blob2,
    unpack_length_prefixed_msgpack,
    to_json_compatible,
)


def configure_parser(p: argparse.ArgumentParser) -> None:
    # No input flags for packs: process response.txt files under decrypt/input recursively
    pass


def run(args: argparse.Namespace) -> int:
    in_root = Path("decrypt/input")
    response_files = list(in_root.rglob("response.txt"))
    if not response_files:
        print("No response.txt files found under decrypt/input")
        return 0

    processed = 0
    for path in sorted(response_files):
        # Match request.txt in the same folder
        req_path = path.with_name("request.txt")
        if not req_path.exists():
            print(f"Skip (no matching request.txt in same folder): {path}")
            continue

        try:
            req_raw = read_base64_file(str(req_path))
            h, _ = parse_request(req_raw)
            req_header = parse_header_blob1(h)
        except Exception:
            print(f"Skip (invalid request.txt format): {req_path}")
            continue

        resp_raw = read_base64_file(str(path))
        udid_str = udid_raw_to_canonical_string(req_header["udid_raw"])
        iv = derive_iv_from_udid_string(udid_str)

        decrypted, key_used = decrypt_blob2(resp_raw, iv)
        payload = unpack_length_prefixed_msgpack(decrypted)
        printable = to_json_compatible(payload)

        rel = path.relative_to(in_root)
        out_dir = Path("decrypt/output") / rel.parent / Path(path).stem
        out_dir.mkdir(parents=True, exist_ok=True)

        bin_path = out_dir / "decoded.bin"
        bin_path.write_bytes(decrypted)

        header_json = {
            "prefix_hex": req_header["prefix"].hex(),
            "prefix_len": len(req_header["prefix"]),
            "session_id_hex": req_header["session_id"].hex(),
            "udid_raw_hex": req_header["udid_raw"].hex(),
            "udid_canonical": udid_str,
            "response_key_hex": req_header["response_key"].hex(),
            "auth_key_hex": req_header["auth_key"].hex(),
            "encryption_key_hex": key_used.hex(),
        }

        combined = {"blob1": header_json, "blob2": printable}
        json_path = out_dir / "decoded.json"
        with open(json_path, "w", encoding="utf-8") as f:
            json.dump(combined, f, ensure_ascii=False, indent=2)

        rel_pack = rel.parent if str(rel.parent) != "." else Path("/")
        print(f"OK: response decrypted -> {json_path}")
        print(f"- Pack: {rel_pack} / {Path(path).stem}")
        print(f"- UDID: {udid_str}")
        print(f"- AES key used: {key_used.hex()}")
        print(f"- Decrypted bin: {bin_path}")
        processed += 1

    if processed == 0:
        print("No valid response.txt files detected under decrypt/input")
    return 0


def main(argv: list[str]) -> int:
    parser = argparse.ArgumentParser(description="Decrypt all packs' response.txt under decrypt/input (recursive)")
    configure_parser(parser)
    args = parser.parse_args(argv)
    return run(args)


if __name__ == "__main__":
    import sys

    raise SystemExit(main(sys.argv[1:]))

File: src/decrypt/response.ts
import path from 'node:path';
import fs from 'node:fs';
import fg from 'fast-glob';
import {
  readBase64File,
  parseRequest,
  parseHeaderBlob1,
  udidRawToCanonicalString,
  deriveIvFromUdidString,
} from '../common/protocol';
import { decryptBlob2, unpackLengthPrefixedMsgpack, toJsonCompatible } from './common';

export async function run(): Promise<number> {
  const inRoot = 'decrypt/input';
  const entries = await fg('**/response.txt', { cwd: inRoot, dot: false });
  if (entries.length === 0) {
    console.log('No response.txt files found under decrypt/input');
    return 0;
  }
  let processed = 0;
  for (const relPath of entries.sort()) {
    const fullPath = path.join(inRoot, relPath);
    const reqPath = path.join(path.dirname(fullPath), 'request.txt');
    if (!fs.existsSync(reqPath)) {
      console.log(`Skip (no matching request.txt in same folder): ${fullPath}`);
      continue;
    }
    try {
      const reqRaw = readBase64File(reqPath);
      const [reqBlob1] = parseRequest(reqRaw);
      const reqHeader = parseHeaderBlob1(reqBlob1);
      const respRaw = readBase64File(fullPath);
      const udidStr = udidRawToCanonicalString(reqHeader.udid_raw);
      const iv = deriveIvFromUdidString(udidStr);
      const { plaintext, keyUsed } = decryptBlob2(respRaw, iv);
      const payload = unpackLengthPrefixedMsgpack(plaintext);
      const printable = toJsonCompatible(payload);

      const outDir = path.join('decrypt/output', path.dirname(relPath), path.parse(fullPath).name);
      fs.mkdirSync(outDir, { recursive: true });
      fs.writeFileSync(path.join(outDir, 'decoded.bin'), plaintext);
      const headerJson = {
        prefix_hex: reqHeader.prefix.toString('hex'),
        prefix_len: reqHeader.prefix.length,
        session_id_hex: reqHeader.session_id.toString('hex'),
        udid_raw_hex: reqHeader.udid_raw.toString('hex'),
        udid_canonical: udidStr,
        response_key_hex: reqHeader.response_key.toString('hex'),
        auth_key_hex: reqHeader.auth_key.toString('hex'),
        encryption_key_hex: keyUsed.toString('hex'),
      };
      const combined = { blob1: headerJson, blob2: printable };
      const jsonPath = path.join(outDir, 'decoded.json');
      fs.writeFileSync(jsonPath, JSON.stringify(combined, null, 2), {
        encoding: 'utf-8',
      });

      console.log(`OK: response decrypted -> ${jsonPath}`);
      const pack = path.dirname(relPath) || '/';
      console.log(`- Pack: ${pack} / response`);
      console.log(`- UDID: ${udidStr}`);
      console.log(`- AES key used: ${keyUsed.toString('hex')}`);
      console.log(`- Decrypted bin: ${path.join(outDir, 'decoded.bin')}`);
      processed += 1;
    } catch (_err) {
      console.log(`Skip (invalid input format): ${fullPath}`);
      continue;
    }
  }
  if (processed === 0) console.log('No valid response.txt files detected under decrypt/input');
  return 0;
}
